### PR TITLE
Test driven intro

### PR DESCRIPTION
**Pull request by kathleen-lund**
Test-driven development exercise. Wrote query() function to return time ranges available for a meeting, first with just mandatory attendees. Then, wrote tests for supporting mandatory and optional employees. Then, amended the query() function to pass those new tests. If there are time ranges where optional attendees can make it along with mandatory, then only those ranges are returned. Otherwise, all times where mandatory attendees can make it are returned.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -34,6 +34,7 @@
   // Some people that we can use in our tests.
   private static final String PERSON_A = "Person A";
   private static final String PERSON_B = "Person B";
+  private static final String PERSON_C = "Person C";
```
Comment (tzsebe) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:37]: Just a heads-up note: Since we're focusing on TDD, I'm reading this file first. As I go through and leave comments, I haven't read the implementation yet, and don't plan to change my comments after I go and read them, even if it clarifies questions.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 2]
```diff
@@ -270,5 +274,151 @@ public void notEnoughRoom() {
...
     Assert.assertEquals(expected, actual);
   }
-}
 
+  @Test
+  public void everyAttendeeIsConsideredWithOptional() {
+    // Have each person have different events. We should see two options because each person has
+    // split the restricted times.
+    //
+    // Events  :       |--A--|     |--B--|
```
Comment (tzsebe) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Nit: It looks like you're trying to encode "who's optional in the new meeting"; it might actually make it even easier to understand if you just had as many rows as you need for accepted calendar attendees (maybe even one row per person so you can see the individual calendars), and then separately list the guests, and optional attendees.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 3]
```diff
@@ -270,5 +274,151 @@ public void notEnoughRoom() {
...
+        new Event("Event 3", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true),
+            Arrays.asList(PERSON_C)));
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
```
Comment (tzsebe) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:319]: Nit: It might be more readable if the first result were on the next line, so everything's pleasantly vertically aligned.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 4]
```diff
@@ -270,5 +274,151 @@ public void notEnoughRoom() {
...
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void everyAttendeeIsConsideredWithOptionalIncluded() {
+    // Have each person have different events. We should see two options because each person has
```
Comment (tzsebe) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:328]: This test case confused me at first, until I realized that if there are times where everyone can attend, then those are the ones that are displayed.

This leads me to ask myself: "What happens if there are multiple optional attendees? Should we not be returning all time slots that all required attendees can make, just sorted by how many optional attendees can make it?" (e.g. in this case, shouldn't the output be [1, 3, 2] where 2 is the time slot covering C?)

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 5]
```diff
@@ -270,5 +274,151 @@ public void notEnoughRoom() {
...
+            Arrays.asList(PERSON_A, PERSON_B)));
+
+    MeetingRequest request = new MeetingRequest(Arrays.asList(), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES), 
+            TimeRange.fromStartEnd(TIME_1000AM, TIME_1200PM, false),
```
Comment (tzsebe) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:420]: If the request is for 30 minute chunks, is it reasonable to return a 2 hour block? (disregard if that's in the assignment spec).

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -270,5 +274,151 @@ public void notEnoughRoom() {
...
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList();
+
+    Assert.assertEquals(expected, actual);
+  }
+}
```
Comment (tzsebe) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:450]: Really good job with this!

Some test cases you should consider adding:

1. Degenerate/edge/corner cases
    1. Request time with no attendees
    1. Add as "optional" a person who is not in the events list (should probably throw a checked exception)
1. A case where there are NO gaps that are free for all required attendees.
1. Basic/Easy cases - All the basic examples, with a single person's calendar events (though I guess the "person" doesn't even matter unless you're worrying about 
1. What if there are no gaps where all people can attend, but a few gaps where various different optional invitees can't make it? Do you prioritize based on most number of people who can make it?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -15,9 +15,181 @@
...
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  public static final int START_OF_DAY = 0;
+  public static final int END_OF_DAY = 1440;
```
Comment (tzsebe) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: When naming variables that represent a unit of time like this, it's best to include the unit in the name (e.g. END_OF_DAY_MINUTES). I'd also add a comment saying this is for a 24 hour day.

Also, for extra readability, it's customary to do the multiplication, so this could be:

`public static final int END_OF_DAY_MINUTES = 24 * 60;`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -15,9 +15,181 @@
...
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Sort Collection by start time by using a stream
```
Comment (tzsebe) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: I'd remove the "using a stream" part of this comment.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -15,9 +15,181 @@
...
+    coalesceList(optionalUnavailableTimes);
+
+    List<TimeRange> availableTimes = new ArrayList<>();
+    List<TimeRange> optionalAvailableTimes = new ArrayList<>();
+    long duration = request.getDuration();
+
+    if (!unavailableTimes.isEmpty()) {
+      // Check if any times are open for mandatory attendees
+      // Methods add open times to availableTimes list
+      checkMorning(unavailableTimes, availableTimes, duration);
```
Comment (tzsebe) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit: Since the only difference is that we're comparing against the day's start time, rather than some other event's end time, I wonder if this is kind of redundant and could be done with less code later.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -15,9 +15,181 @@
...
+      return totalAvailableTimes;
+    }    
+  }
+
+ /**
+  * coalesceList: runs through a list of times and coalesces any
+  * overlapping times so that every entry in the list is a 
+  * distinct and separate time.
+  */
+  public static void coalesceList(List<TimeRange> times) {
```
Comment (tzsebe) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I would make this private; it's not really relevant on its own to the functionality of the class its in (aside from being an internal helper), and it's also pretty specialized, since it changes the list in place.

If this were to be a public interface, I'd move it out of this class into a separate utility class, and then I'd have it assert that the list is correctly sorted (and fail if it's not). I'd also likely have it create a copy, rather than change the list in place. THAT SAID, don't do this - I'm just saying what it'd take for it to warrant being public.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -15,9 +15,181 @@
...
+    }
+    else if (!optionalAttendees.isEmpty() && duration <= END_OF_DAY) {
+      // Entire day is open
+      optionalAvailableTimes.add(TimeRange.fromStartEnd(START_OF_DAY, END_OF_DAY, false));
+    }
+
+    // Attempt to combine to two lists to see if there are times when mandatory
+    // and optional attendees are both available
+    List<TimeRange> totalAvailableTimes = combineAvailableLists(availableTimes, optionalAvailableTimes, duration);
+
```
Comment (tzsebe) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I think you might be complicating your life a bit in your implementation of this method. I had a hard time following the if block below, because of the interplay of all the artifacts you generate above.

Consider trying some logic that looks more like:

```
// Just pseudo-code to show the logic flow... and you'd have to implement invert.
// invert is just "all time windows NOT in the sorted event lists" - note that you don't
// need to coalesce if they're sorted... just follow the chain until there's a gap between two events.
allAvailableTimes = filterTimeWindowLength(invert(allEvents));
allRequiredAvailableTimes = filterTimeWindowLength(invert(allRequiredEvents));
return allAvailableTimes.isEmpty()
    ? allRequiredAvailableTimes
    : allAvailableTimes;
```
(unless you want to get into fancy sorting by "number of people who can't make it" or somesuch later)

**Review by tzsebe — CHANGES_REQUESTED** [review]
I really love the graphical way you're documenting what these tests do. This makes it much easier to understand what you're testing for. A++ :)

I hope you enjoyed TDD! You'll especially appreciate that you'll be able to do the refactor I'm proposing, and you can be confident you're doing it correctly, because you already have a test corpus to validate it! :)

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 2 above)
Comment (kathleen-lund) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Thanks so much for pointing this out--I like that I was able to change this and I think it makes a lot more sense now.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 4 above)
Comment (kathleen-lund) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:328]: Yes, so the default behavior requested was to only return times all mandatory/optional could make it or, if there's none of those, return times all mandatory could make it. Sorting by number of optional employees that can make it was one of the extra "challenge activities" for the week I mentioned.  The changes I just pushed don't include that functionality, but I think I may have some time to work on it the rest of today and tomorrow so I may make a new branch for it if that sounds okay?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 5 above)
Comment (kathleen-lund) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:420]: Yes, it was the assignment spec to just return entire chunks that were long enough.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 6 above)
Comment (kathleen-lund) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:450]: Thank you!!
As for other tests, the starter code for the assignment actually came with a bunch of tests already done. (This also means that I can't take credit for the awesome visual comments you mentioned above--the template for them was there when I arrived :)) I think 1i, 2, and 3 were included in the tests we started with that I didn't modify. 
I've got a question about 1ii: if there's an optional person who isn't in the events list, does that just mean their whole day is open?
And for 4, I'll add this if I have time to make that new branch/functionality as I described above, if that sound good!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 12]
```diff
@@ -99,9 +104,11 @@ public void everyAttendeeIsConsidered() {
     // Have each person have different events. We should see two options because each person has
     // split the restricted times.
     //
-    // Events  :       |--A--|     |--B--|
-    // Day     : |-----------------------------|
-    // Options : |--1--|     |--2--|     |--3--|
+    // Mandatory: A, B
+    // Events   :       |--A--|
```
Comment (mkmg) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:108]: Nit: Call this "Person A's Events" and the below one "Person B's Events" to make it a bit clearer.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 13]
```diff
@@ -99,9 +104,11 @@ public void everyAttendeeIsConsidered() {
     // Have each person have different events. We should see two options because each person has
```
Comment (mkmg) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:104]: It says two options here, but the diagram shows three options. Is this "two" a typo?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 11 above)
Comment (kathleen-lund) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This implementation makes so much more sense; thank you! I've implemented it in the latest commit.
I still ended up having to coalesce the lists of times where there were events, because if I had an all-day event for someone with two smaller events contained within it, for example, the invert method would end up adding one or more of the gaps between the smaller events as I was following the chain of the 3+ events. It worked once I coalesced everything together, but please let me know if there was something I could have done differently in invert that would have eliminated the need for coalesce, because I could not think of a way.
Thank you again for your comments and helping me really clean up the code!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 14]
```diff
@@ -270,5 +287,164 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
-}
 
+  @Test
+  public void everyAttendeeIsConsideredWithOptional() {
+    // Have each person have different events. We should see two options because each person has
```
Comment (mkmg) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:293]: Same question about the "two" vs "three" options here.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 15]
```diff
@@ -15,9 +15,137 @@
...
+    List<TimeRange> allRequiredEvents = new ArrayList<>();
+    for (Event e : events) {
+      // If the attendee lists are not disjoint (meaning there is at least one attendee in common), 
+      // save this event time
+      if (!Collections.disjoint(e.getAttendees(), mandatoryAttendees)) {
+        // Required for everyone
+        allRequiredEvents.add(e.getWhen());
+        allEvents.add(e.getWhen());
+      } 
+      else if (!Collections.disjoint(e.getAttendees(), optionalAttendees)) {
```
Comment (mkmg) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:51]: No new line between '}' and 'else if'.

https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style

**Review by mkmg — APPROVED** [review]
You did a really good job breaking the algorithm down into nice helper methods. Great work!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 6 above)
Comment (tzsebe) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:450]: For 1ii: It's ambiguous! You make the judgment call hehe. Since this is a real thing that can happen, it's worth actually specifying it through tests and in the docs, since users of this code might have the same question, and might even build something the assumes one behavior or the other.

Comment (tzsebe) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:450]: For 4 - if you like; again, it's not a priority, only do it if you're learning useful things by doing it. If it's just busy work, then focus on the capstone.

**Review by tzsebe — APPROVED** [review]
Awesome, excellent work!
